### DISCUSSION TITLE
Unable to run MongoDb ReplicaSet

### DISCUSSION THREAD

**Discussion by diegosasw** [testcontainers/testcontainers-dotnet#1150] [Q&A] [answered]
I have successfully used Ductus.FluentDocker to spin up a mongoDb container with a replica set. This is how I achieved it.

Notice it's in 2 steps, first I start the container with the `mongod` command, and later I initialize the replicaSet with `mongosh` instruction.

```
const string replicaSetHost = "mongo";

var mongoDbService =
	new Builder()
		.UseContainer()
		.WithName(replicaSetHost)
		.UseImage("mongo:7.0.7-jammy")
		.ExposePort(27017, 27017)
		.Command("mongod", "--replSet", "rs0", "--bind_ip_all")
		.DeleteIfExists()
		.RemoveVolumesOnDispose()
		.UseNetwork(network)
		.Build();

mongoDbService.Start();

var hosts = new Hosts().Discover();
var host = hosts.FirstOrDefault(h => h.IsNative) ?? hosts.FirstOrDefault(h => h.Name == "default");
var runningContainers = host.GetRunningContainers();
// Iterate over running containers to find the mongo one
foreach (var runningContainer in runningContainers)
{
	var name = runningContainer.Name;
	var isMongoContainer = replicaSetHost == name;
	if (isMongoContainer)
	{
		var mongoInitiateCommand = 
			$@"mongosh --eval ""rs.initiate({{_id: 'rs0', members: [{{_id: 0, host: '{replicaSetHost}:27017'}}]}});""";

		var responseInitiate = runningContainer.Execute(mongoInitiateCommand);
		if (!responseInitiate.Success)
		{
			throw new Exception("Could not initiate replica set in Mongo");
		}

		const string mongoStatusCommand = "mongosh --eval \"rs.status();\"";
		var responseConfig = runningContainer.Execute(mongoStatusCommand);
		if (!responseConfig.Success)
		{
			throw new Exception("Could not check replica set status in Mongo");
		}
	}
}
```

This will spin up a container, and I can access it with connection string `mongodb://mongo:27017/?replicaSet=rs0&ssl=false`

Notice the hostname is `mongo` (it could be potato, it doesn't matter) and I have an entry on my /etc/hosts file so that I can use the alias mongo as if it was localhost. That's because initiating the replica set expects the host name.
```
127.0.0.1 mongo
```

Now the problem is that I would like to achieve the same with TestContainers for MongoDb and I cannot make it work.

The following code gets stuck on starting the container

```
var mongoDb =
      new MongoDbBuilder()
           .WithImage("mongo:7.0.7-jammy")
           //.WithEntrypoint("mongod")
           .WithCommand("mongod", "--replSet", replicaSetName, "--bind_ip_all")
           .Build();

await mongoDb.StartAsync();
```

and if I uncomment the entrypoint part, it throws a timeout exception.

Has anyone managed to have a replica set working with testcontainers? Ideally a single node replica set, which is the minimum necessary for using change streams.

UPDATE:
I found out, if I pass empty username and empty password, it does not get stuck there. But, still, I am unable to have it working

This is the furthest I got, and it fails to initialize the replica set, although the mongoDb container spins up.
```
const string replicaSetName = "rs0";
const string hostname = "mongo";

var mongoContainer =
    new MongoDbBuilder()
        .WithUsername("") // it needs to be empty
        .WithPassword("") // it needs to be empty
        .WithImage("mongo:7.0.7-jammy")
        .WithHostname(hostname)
        // do I really need .WithExtraHost() ?
        .WithCommand("mongod", "--replSet", replicaSetName, "--bind_ip_all")
        .WithWaitStrategy(Wait.ForUnixContainer())
        .Build();

await mongoContainer.StartAsync();
var port = mongoContainer.GetMappedPublicPort(27017);
var mongoInitiateCommand = 
      $@"mongosh --eval ""rs.initiate({{_id: '{replicaSetName}', members: [{{_id: 0, host: '{hostname}:{port}'}}]}});""";
var result = await mongoContainer.ExecAsync([mongoInitiateCommand]);
var connectionString = mongoContainer.GetConnectionString() ??  $"mongodb://{hostname}:{port}/?replicaSet=rs0&ssl=false";
```

**Comment by diegosasw** [comment]
To add an additional note on this, If I were to spin up a MongoDb replica set using Docker/Docker compose, I would do the following.

Run a `docker-compose up -f mongo-rs.yml` for this `mongo-rs.yml`:

```
version: '3.8'

services:
  mongodb:
    image: mongo:latest
    command: ["mongod", "--replSet", "rs0", "--bind_ip_all"]
    ports:
      - "27017:27017"
```

Find out the container Id with a `docker ps`, for example.

And then connect to the container's mongosh
```
docker exec -it <container_id> mongosh
```
and run the following
```
test> const configuration = { _id: "rs0", members: [{ _id: 0, host: "127.0.0.1:27017"}]}

test> rs.initiate(configuration)
```

so that it is reachable at `mongodb://127.0.0.1:27017/?replicaSet=rs0` or whichever host name alias is specified.

I would like the same with testcontainer. Has anyone succeeded?

PS: I found this StackOverflow question and I added my findings https://stackoverflow.com/a/78277081/2948212

**Comment by HofmeisterAn** [comment]
Hi 👋 If a module does not support a feature, it is usually a good idea to look into Java's implementation. Some modules offer more features than their .NET counterparts and vice versa. In this case, AFAIK, Java supports [replica sets](https://github.com/testcontainers/testcontainers-java/blob/9d6e7655a929abcc4db735e16642e4eb40e12155/modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java#L171). I think for proper support, it would be good to align the implementation and add the missing pieces from Java to .NET.

**Comment by HofmeisterAn** [comment, marked as answer]
See:

- https://github.com/testcontainers/testcontainers-dotnet/issues/1154
- https://github.com/testcontainers/testcontainers-dotnet/pull/1196

Will be included in the next release 3.10.0.
